ads/google/a4a/traffic-experiments.js

```diff
@@ -22,12 +22,100 @@
+import {dev} from '../../../src/log';
+import {getMode} from '../../../src/mode';
 
-/** @typedef {{string: {branches: {control: string, experiment: string}}}} */
+/** @typedef {{string: {branches: !Branches}}} */
+export let Branches;
+
+/** @typedef {{control: string, experiment: string}} */
 export let ExperimentInfo;
 
+/**
+ * Check whether Google Ads supports the A4A rendering pathway for a given ad
+ * Element on a given Window.  The tests we use are:
+ * <ol>
+ *   <li>The page must have originated in the {@code cdn.ampproject.org}
+ *   CDN <em>or</em> we must be running in local dev mode.</li>
+ *   <li>We must be selected in to an A4A traffic experiment and be selected
+ *   into the "experiment" branch.
+ * </ol>
+ *
+ * If we're selected into the overall traffic experiment, this function will
+ * also attach an experiment or control branch ID to the {@code Element} as
+ * a side-effect.
+ *
+ * @param {!Window} win  Host window for the ad.
+ * @param {!Element} element Ad tag Element.
+ * @param {string} experimentId ID for the experiment to use.
+ * @param {!Branches} experiment and control branch IDs to use.
+ * @return {boolean}  Whether Google Ads should attempt to render via the A4A
+ *   pathway.
+ */
+export function googleAdsIsA4AEnabled(win, element, experimentId, branches) {
+  // Note: Theoretically, isProxyOrigin is the right way to do this, b/c it
+  // will be kept up to date with known proxies.  However, it doesn't seem to
+  // be compatible with loading the example files from localhost.  To hack
+  // around that, just say that we're A4A eligible if we're in local dev
+  // mode, regardless of origin path.
+  if (isGoogleAdsA4AValidEnvironment(win)) {
+    // Page is served from a supported domain.
+    handleUrlParameters(win, experimentId, branches);
+    setupPageExperiments(win, {[experimentId]: branches});
+    if (isExperimentOn(win, experimentId)) {
+      // Page is selected into the overall traffic experiment.
+      if (getPageExperimentBranch(win, experimentId) === branches.experiment) {
+        // Page is on the "experiment" (i.e., use A4A rendering pathway)
+        // branch of the overall traffic experiment.
+        addExperimentIdToElement(branches.experiment, element);
+        return true;
+      } else {
+        // Page is on the "control" (i.e., use traditional, 3p iframe
+        // rendering pathway) branch of the overall traffic experiment.
+        addExperimentIdToElement(branches.control, element);
+        return false;
+      }
+    }
+  }
+  // Serving location doesn't qualify for A4A treatment or page is not in the
+  // traffic experiment.
+  return false;
+}
+
+/**
+ * Set experiment state from URL parameter, if present.
+ *
+ * @param {!Window} win
+ * @param {string} experimentId
+ * @param {!Branches} branches
+ * @visibleForTesting
+ */
+function handleUrlParameters(win, experimentId, branches) {
+  const a4aParam = /(?:\?|&)a4a=([0-9]+)/.exec(win.location.search);
```

Comment:
We discussed it.  Decided it wasn't worth parsing the URL fully in order to pull out one parameter.  If you feel that readability / code reuse dominates, we can switch it.